Std/Lean/ToExpr.lean

```diff
@@ -0,0 +1,82 @@
+/-
+Copyright (c) 2023 Kyle Miller. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kyle Miller
+-/
+import Std.Lean.Elab.Deriving.ToExpr
+
+/-! # `ToExpr` instances
+
+This module should be imported by any module that intends to define `ToExpr` instances.
+It provides necessary dependencies (the `Lean.ToLevel` class) and it also overrides the instances
+that come from core Lean 4 that do not handle universe polymorphism.
+(See the module `Lean.ToExpr` for the instances that are overridden.)
+
+In addition, we provide some additional `ToExpr` instances for core definitions.
+-/
+
+namespace Lean
+
+attribute [-instance] Lean.instToExprOption
+
+deriving instance ToExpr for Option
+
+attribute [-instance] Lean.instToExprList
+
+deriving instance ToExpr for List
+
+attribute [-instance] Lean.instToExprArray
+
+instance {α : Type u} [ToExpr α] [ToLevel.{u}] : ToExpr (Array α) :=
+  let type := toTypeExpr α
+  { toExpr     := fun as => mkApp2 (mkConst ``List.toArray [toLevel.{u}]) type (toExpr as.toList)
+    toTypeExpr := mkApp (mkConst ``Array [toLevel.{u}]) type }
+
+attribute [-instance] Lean.instToExprProd
+
+deriving instance ToExpr for Prod
+
+deriving instance ToExpr for System.FilePath
+
+deriving instance ToExpr for Int
+
+deriving instance ToExpr for ULift
+
+/-- Hand-written instance since `PUnit` is a `Sort` rather than a `Type`. -/
+instance [ToLevel.{u}] : ToExpr PUnit.{u+1} where
+  toExpr _ := mkConst ``PUnit.unit [toLevel.{u+1}]
+  toTypeExpr := mkConst ``PUnit [toLevel.{u+1}]
+
+deriving instance ToExpr for String.Pos
+deriving instance ToExpr for Substring
+deriving instance ToExpr for SourceInfo
+deriving instance ToExpr for Syntax.Preresolved
+deriving instance ToExpr for Syntax
+
+open DataValue in
+private def toExprMData (md : MData) : Expr := Id.run do
```

Comment:
```
/-- Core of a hand-written `ToExpr` handler for `MData`.
Uses the `KVMap.set*` functions rather than going into the internals
of the `KVMap` data structure. -/
```

Reply:
```suggestion
/-- Core of a hand-written `ToExpr` handler for `MData`.
Uses the `KVMap.set*` functions rather than going into the internals
of the `KVMap` data structure. -/
private def toExprMData (md : MData) : Expr := Id.run do
```